[http1 codec] Allow requests with Transfer-Encoding and Content-Length headers set.

veshij opened the pull request:
Commit Message: [http1 codec] Allow HTTP/1 requests/responses with Transfer-Encoding and Content-Length headers set.
Additional Description: opt-in for serving requests/responses with Content-Length and Transfer-Encoding: chunked. Per RFC remove Content-Length header before forwarding it to upstream.
Update http-parser version to include fix for https://github.com/nodejs/http-parser/issues/517.
Risk Level: Medium
Testing: unit tests, update integration test
Docs Changes: updated docs/root/version_history/current.rst
Release Notes: http: added allow_chunked_length configuration option for HTTP/1 codec to allow processing requests with both Content-Length and Transfer-Encoding: chunked. If enabled - per RFC Content-Length is removed before sending to upstream.

Fixes #11398

alyssawilk commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -459,6 +459,7 @@ ConnectionImpl::ConnectionImpl(Network::Connection& connection, CodecStats& stat
       max_headers_kb_(max_headers_kb), max_headers_count_(max_headers_count) {
   output_buffer_.setWatermarks(connection.bufferLimit());
   http_parser_init(&parser_, type);
+  parser_.allow_chunked_length = 1;
```
Given the danger of smuggling I'd be inclined to disable this by default, and allow it only iff it's configured on via API or runtime hooks.  
@mattklein123 @antoniovicente for second opinion.

If they rule I'm being overly concerned, we'll instead want to runtime guard as it's a behavioral change, and either way we will want release notes and tests of the new behavior :-)

mattklein123 replied:
+1. I think this should be a config option that is set on the HTTP protocol options? That way it can be set independently for upstream and downstream? WDYT? It shouldn't be hard to add this as the options are already plumbed through everywhere IIRC.

veshij replied:
The idea was to:
 * enable non-safe parser configuration by default in envoy
 * enable strict validation by default in envoy (so we don't rely on parser behavior)
and on top of that
* add configuration option (http protocol options) to disable strict validation in envoy.

I can update diff to enable non-safe parser mode only if configuration option is set. If so - should I leave sanity checking enabled by default (even if we know default parser won't allow such request)?

mattklein123 replied:
I think it's fine to do what you propose (on all the time in parser, block in Envoy code), but will defer to @alyssawilk if she has a strong opinion on it.

alyssawilk replied:
Yeah I'm fine with that, as long as we double check our tests as we go to make sure we're actually doing the blocking in envoy.

alyssawilk commented on source/common/http/http1/codec_impl.h:
```diff
@@ -201,6 +201,7 @@ class ConnectionImpl : public virtual Connection, protected Logger::Loggable<Log
   }
   uint32_t bufferLimit() { return connection_.bufferLimit(); }
   virtual bool supportsHttp10() { return false; }
+  virtual bool allowChunkedLength() { return false; }
```
Any reason to allow this in one direction and not the other?

mattklein123 replied:
+1 why not just always do this and/or pass the config option in? We have it available in both upstream and downstream?

alyssawilk commented on bazel/repository_locations.bzl:
```diff
@@ -283,9 +283,9 @@ DEPENDENCY_REPOSITORIES = dict(
-        sha256 = "8fa0ab8770fd8425a9b431fdbf91623c4d7a9cdb842b9339289bd2b0b01b0d3d",
-        strip_prefix = "http-parser-2.9.3",
-        urls = ["https://github.com/nodejs/http-parser/archive/v2.9.3.tar.gz"],
+        sha256 = "6a12896313ce1ca630cf516a0ee43a79b5f13f5a5d8143f56560ac0b21c98fac",
```
totally lazy ask - mind replying with the diff links so reviewers can easily see what has changed in http_parser between the two hashes?

mattklein123 replied:
Also please add a comment with the date and the reason for not using a fixed version. See other examples in this file.

veshij replied:
http_parser aggregated change: https://github.com/nodejs/http-parser/compare/a0c034c0c7698c08f8dc8c8d0257305f6280c27b..4f15b7d510dc7c6361a26a7c6d2f7c3a17f8d878

mattklein123 commented on include/envoy/http/codec.h:
```diff
@@ -376,6 +376,10 @@ struct Http1Settings {
   //  - Not a HEAD request
   bool enable_trailers_{false};
+  // Allow serving requests with both Content-Length and Transfer-Encoding: chunked.
+  // In enabled and such request served - Content-Length is ignored and removed from request
```
```suggestion
  // If enabled and such a request is served - Content-Length is ignored and removed from the request
```

mattklein123 commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -878,6 +879,30 @@ int ServerConnectionImpl::onHeadersComplete() {
+      if ((parser_.flags & F_CHUNKED) && allowChunkedLength()) {
+        headers->removeContentLength();
+      } else {
+        ENVOY_CONN_LOG(error, "Both 'Content-Length' and 'Transfer-Encdoding' are set.",
```
We need to log at debug level for data plane logs. I would add a stat if you want to see this case explicitly.

veshij replied:
oops, sure, will fix it.
(is there a way to enable debug logging when running integration test?)

mattklein123 replied:
`--test_arg="-l debug"`

antoniovicente commented:
/assign @antoniovicente

veshij commented:
/wait

veshij commented:
/wait

veshij commented:
/wait

mattklein123 commented on api/envoy/config/core/v3/protocol.proto:
```diff
@@ -157,6 +157,12 @@ message Http1ProtocolOptions {
   //   - The content length header is not present.
   bool enable_trailers = 5;
+
+  // Allows Envoy to process requests with `Content-Length` and `Transfer-Encoding: chunked`.
+  // By default such requests are rejected with `400 Bad Request`, but if option is enabled -
+  // Envoy will remove Content-Length header and pass request to upstream.
+  // See `RFC7230, sec. 3.3.3 <https://tools.ietf.org/html/rfc7230#section-3.3.3>` for details.
```
Can you add an `.. attention:` block here that warns about request smuggling?

mattklein123 commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -878,6 +879,30 @@ int ServerConnectionImpl::onHeadersComplete() {
+    // perform request smuggling (Section 9.5) or response splitting
+    // (Section 9.4) and ought to be handled as an error. A sender MUST
+    // remove the received Content-Length field prior to forwarding such
+    // a message downstream.
```
```suggestion
    // a message.
```

mattklein123 commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -878,6 +879,30 @@ int ServerConnectionImpl::onHeadersComplete() {
+
+    // Reject request with Http::Code::BadRequest by default or remove Content-Length header
+    // and serve request if allowed by http1 codec settings.
+    if (parser_.uses_transfer_encoding != 0 &&
```
Sorry why this uses_transfer_encoding check? Couldn't you just do `(parser_.flags & F_CHUNKED)`? Or you are checking for any transfer encoding? Can you add more comments?

veshij replied:
I was following RFC here (any TE, not only `chunked`):
```
If a message is received with both a Transfer-Encoding and a
Content-Length header field, the Transfer-Encoding overrides the
Content-Length.
```

mattklein123 commented on test/integration/integration_test.cc:
```diff
@@ -425,7 +425,8 @@ TEST_P(IntegrationTest, TestSmuggling) {
                                 smuggled_request;
     sendRawHttpAndWaitForResponse(lookupPort("http"), request.c_str(), &response, false);
-    EXPECT_THAT(response, HasSubstr("HTTP/1.1 400 Bad Request\r\n"));
+    // 'identity' encoding is not supported
+    EXPECT_THAT(response, HasSubstr("HTTP/1.1 501 Not Implemented\r\n"));
```
This seems not right. Why are we now returning a 501 here?

veshij replied:
onHeadersCompleteBase checks for unsupported encodings (`identity` in this case) and returns 501 if unsupported encoding is detected.
if everything is ok - passes control to onHeadersComplete() where my code is executed.

I can refactor it to do CL/TE checks before TE encoding check which returns 501, I'm not sure which order is correct in this case.

veshij replied:
nginx's response in case of unsupported encoding and CL set:
```
> POST / HTTP/1.1
> Host: example.com
> User-Agent: curl/7.64.1
> Accept: */*
> Transfer-Encoding: unsupported
> Content-Length: 10
>
< HTTP/1.1 501 Not Implemented
< Server: nginx
< Date: Fri, 31 Jul 2020 20:40:47 GMT
< Content-Type: text/html
< Content-Length: 14480
< Connection: close
```

mattklein123 replied:
Naively it seems like we should return a 4xx code here, but I'm not an expert. If this is what NGINX does that seems fine, thanks.

mattklein123 replied:
OK I see by allowing this in the parser we get slightly different behavior here. I think this is OK but I think we should release note this also and I would also like to hear from @antoniovicente or @PiotrSikora on this.

mattklein123 replied:
Friendly ping on the release note on the subtle behavior change here.

/wait

antoniovicente replied:
4xx makes more sense to me in this case.

This specific change seems unrelated to the rest of the functional changes in this PR, I would suggest doing the http_parser version bump separately from the rest of this PR.

veshij replied:
I can split this PR in two, one for library update and one for functional change.

mattklein123 replied:
> This specific change seems unrelated to the rest of the functional changes in this PR, I would suggest doing the http_parser version bump separately from the rest of this PR.

@veshij correct me if I'm wrong, but I don't think this has to do with the library update. It has to do with the fact that we are now hitting code for this case that we did not hit before, and Envoy is returning 501 when it used to return 400. Right? If so I think this seems OK to me with a release note but will defer to @antoniovicente and @PiotrSikora on that.

veshij replied:
>I would suggest doing the http_parser version bump separately
Sorry, that was a reply to comment above.

mattklein123 commented on source/common/http/http1/codec_impl_legacy.h:
```diff
@@ -205,6 +205,7 @@ class ConnectionImpl : public virtual Connection, protected Logger::Loggable<Log
   }
   uint32_t bufferLimit() { return connection_.bufferLimit(); }
   virtual bool supportsHttp10() { return false; }
+  virtual bool allowChunkedLength() { return false; }
```
This part needs to be synced up. (Sorry for the trouble here while we converge this).

antoniovicente commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -459,6 +459,7 @@ ConnectionImpl::ConnectionImpl(Network::Connection& connection, CodecStats& stat
       max_headers_kb_(max_headers_kb), max_headers_count_(max_headers_count) {
   output_buffer_.setWatermarks(connection.bufferLimit());
   http_parser_init(&parser_, type);
+  parser_.allow_chunked_length = 1;
```
What about the case of responses?  It seems like the code to respect the config is specific to the server side, but the option to allow_chunked_length is set unconditionally on the base class which is shared across request and response processing.

I only see new tests for the request case.  Does the response case have appropriate coverage?

PiotrSikora replied:
+1.

mattklein123 replied:
Ah yes good catch. We need to do this for responses also and either raise a protocol error or do the translation.

veshij replied:
Added response checks, still working on tests.

antoniovicente commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -878,6 +879,30 @@ int ServerConnectionImpl::onHeadersComplete() {
+    // Reject request with Http::Code::BadRequest by default or remove Content-Length header
+    // and serve request if allowed by http1 codec settings.
+    if (parser_.uses_transfer_encoding != 0 &&
+        (parser_.content_length > 0 && parser_.content_length != ULLONG_MAX)) {
```
what is the parser_.content_length != ULLONG_MAX about?

Also, is this case handled correctly?
transfer-encoding: chunked
content-length: 0

Please add a test with content-length == 0.

veshij replied:
Updated code, added testcases with different Content-Length.

PiotrSikora commented on api/envoy/config/core/v3/protocol.proto:
```diff
@@ -157,6 +157,16 @@ message Http1ProtocolOptions {
   bool enable_trailers = 5;
+
+  // Allows Envoy to process requests with `Content-Length` and `Transfer-Encoding: chunked`.
+  // By default such requests are rejected with `400 Bad Request`, but if option is enabled -
```
I'm not sure if I agree with the default here. Yes, if both are present, then `Transfer-Encoding` takes a precedence, and `Content-Length` may be removed, but rejecting such request seems a bit harsh.

veshij replied:
That's the current behavior (http-parser fails parsing with HPE_UNEXPECTED_CONTENT_LENGTH which ends up as Bad Request to the downstream), I'm keeping it consistent.

mattklein123 commented on docs/root/version_history/current.rst:
```diff
@@ -15,6 +15,7 @@ Minor Behavior Changes
 * http: changed Envoy to send error headers and body when possible. This behavior may be temporarily reverted by setting `envoy.reloadable_features.allow_response_for_timeout` to false.
 * http: clarified and enforced 1xx handling. Multiple 100-continue headers are coalesced when proxying. 1xx headers other than {100, 101} are dropped.
 * http: fixed the 100-continue response path to properly handle upstream failure by sending 5xx responses. This behavior can be temporarily reverted by setting `envoy.reloadable_features.allow_500_after_100` to false.
+* http: http1 codec returns `500 Non Implemented` instead of `400 Bad Request` for request having both Content-Length and Transfer-Encoding headers set and Transfer-Encoding contains unsupported encoding.
```
nit: s/http1/HTTP/1

mattklein123 commented on include/envoy/http/codec.h:
```diff
@@ -374,6 +374,10 @@ struct Http1Settings {
   //  - Is neither a HEAD only request nor a HTTP Upgrade
   //  - Not a HEAD request
   bool enable_trailers_{false};
+  // Allow serving requests with both Content-Length and Transfer-Encoding: chunked.
```
nit: update for response handling also

mattklein123 commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -878,6 +879,28 @@ int ServerConnectionImpl::onHeadersComplete() {
+        error_code_ = Http::Code::BadRequest;
+        sendProtocolError(Http1ResponseCodeDetails::get().ContentLengthNotAllowed);
+        throw CodecProtocolException(
+            "http/1.1 protocol error: both 'Content-Length' and 'Transfer-Encdoding' are set.");
```
typo Encdoding here and below

mattklein123 commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -1115,6 +1139,27 @@ int ClientConnectionImpl::onHeadersComplete() {
+      if ((parser_.flags & F_CHUNKED) && codec_settings_.allow_chunked_length_) {
+        headers->removeContentLength();
+      } else {
+        sendProtocolError(Http1ResponseCodeDetails::get().ContentLengthNotAllowed);
```
I don't think sending protocol error makes sense as a client. All we can do is raise a protocol exception and stat it.

veshij replied:
@mattklein123 I'm a bit confused as other checks above and below this code call sendProtocolError and throw CodecProtocolException

mattklein123 commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -1115,6 +1139,27 @@ int ClientConnectionImpl::onHeadersComplete() {
       }
     }
 
+    // https://tools.ietf.org/html/rfc7230#section-3.3.3
+    // If a message is received with both a Transfer-Encoding and a
+    // Content-Length header field, the Transfer-Encoding overrides the
+    // Content-Length. Such a message might indicate an attempt to
+    // perform request smuggling (Section 9.5) or response splitting
+    // (Section 9.4) and ought to be handled as an error. A sender MUST
+    // remove the received Content-Length field prior to forwarding such
+    // a message.
+
+    // Reject response if Transfer-Encoding and Content-Length headers are present
+    // or remove Content-Length and serve response if allowed by http1 codec settings.
```
This entire block and (and code below) is duplicated in the server case. Can we share a function that checks and handles it? The error case is slightly different but that should be easy to happen once the return status is checked from the shared funtion.

antoniovicente replied:
It may be possible to do this error check in onHeadersCompleteBase.

veshij replied:
moved all validation logic to onHeadersCompleteBase(), code is not duplicated.

mattklein123 commented on source/common/http/http1/codec_impl.h:
```diff
@@ -577,6 +576,7 @@ class ClientConnectionImpl : public ClientConnection, public ConnectionImpl {
     }
   }
 
+  Http1Settings codec_settings_;
```
Move to base class?

veshij commented:
Updated PR:
* validation logic is in onHeadersCompleteBase() which is shared by Client/Server implementations.
* codec_settings_ moved to base class, 
* reverted integration test change and version history comment note as behavior is not changed now if option is off

mattklein123 commented on source/common/http/http1/codec_impl.h:
```diff
@@ -254,7 +254,7 @@ class ConnectionImpl : public virtual Connection, protected Logger::Loggable<Log
   bool deferred_end_stream_headers_ : 1;
   const bool connection_header_sanitization_ : 1;
-  const bool enable_trailers_ : 1;
+  Http1Settings codec_settings_;
```
perf: reorder to avoid splitting the bit field

mattklein123 commented on source/common/http/http1/codec_impl.h:
```diff
@@ -220,7 +220,7 @@ class ConnectionImpl : public virtual Connection, protected Logger::Loggable<Log
   ConnectionImpl(Network::Connection& connection, CodecStats& stats, http_parser_type type,
                  uint32_t max_headers_kb, const uint32_t max_headers_count,
-                 HeaderKeyFormatterPtr&& header_key_formatter, bool enable_trailers);
+                 HeaderKeyFormatterPtr&& header_key_formatter, Http1Settings settings);
```
perf: pass by const ref

antoniovicente commented on api/envoy/config/core/v3/protocol.proto:
```diff
@@ -157,6 +157,16 @@ message Http1ProtocolOptions {
+  // .. attention::
+  //   Enabling this option might lead to request smuggling vulnerability, especially if traffic
+  //   is proxied via multiple layers of proxies.
+  bool allow_chunked_length = 6;
```
Would it be appropriate to add either/both of these annotations:
[
    (udpa.annotations.security).configure_for_untrusted_downstream = true,
    (udpa.annotations.security).configure_for_untrusted_upstream = true
  ];

veshij replied:
I'm not quite sure.
You might want to configure it in case of misbehaving downstream/upstream, but default behavior (reject request/response if both headers are present)  is safe and in general envoy users will not need to change it.
Or am I interpreting these annotations wrong?

antoniovicente replied:
cc @htuch 

I'm not an expert in this area.  At one point someone convinced me that closing the connection in cases where HTTP messages contained content-length and transfer-encoding: chunked was safer than simply removing the content-length.  I think that handling requests with both content-length and transfer-encoding chunked by removing content-length leaves you vulnerable to HTTP de-synchronization attacks if a vulnerable proxy is sitting in front of the Envoy.  Replying with an error and closing the connection may result in safer behavior, but I wouldn't be surprised if de-synchronization were still possible if a vulnerable proxy is sitting in front of the envoy.

RFC 7230 section 3.3.3, bullet point 3 considers this situation to be an error.  It does seem to allow for terminating the request or forwarding it along after removing the content-length header, and those are two options you allow users to select via this config knob.  The most recent revision from 2020 also has similar wording.  I think we are following best practices, I just don't know what level of warning is most appropriate here.

mattklein123 replied:
IMO treating it as an error by default is appropriate and we have the right attention/warning on this to tell people about the risk, but I'm not an expert in this area. @PiotrSikora any thoughts here? @htuch?

PiotrSikora replied:
I think that simply removing `Content-Length` would be fine, but I have no objections if @antoniovicente prefers rejecting requests. Both are valid options.

htuch replied:
From an API perspective, this seems the right choice and we do call attention to the danger of enabling. I'd actually like to understand why this is needed though, is there some real world traffic that needs this? Can we explain this usage in the commit message and/or docs.

veshij replied:
>is there some real world traffic that needs this? 
In our case - some old API clients behave that way. And as nginx ignores Content-Length by default in this case it was unnoticed until we tried to serve this traffic via Envoy.

htuch replied:
Seems legit; even if the old API clients are not RFC conformant, I don't see an issue allowing opt-in. You might want to add to https://www.envoyproxy.io/docs/envoy/latest/configuration/best_practices/edge#best-practices-edge a section on "what not to configure" and include this.

antoniovicente replied:
I think we're doing the right thing by adding the config knob and setting its default to reject.  I think that the remaining question was what to include in the warning and if we wanted to set some security annotations like configure_for_untrusted_downstream on this field.

I think that the comment you added is sufficient.  I'm fine not adding extra security annotations to the config field.  We can always revisit later.

htuch replied:
Yeah, the annotations are for fields which should be _set_ for untrusted downstream. In this case, we have a field that should not be set for untrusted downstream, so I think it's fine to skip.

antoniovicente commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -674,6 +675,29 @@ int ConnectionImpl::onHeadersCompleteBase() {
+      error_code_ = Http::Code::BadRequest;
+      sendProtocolError(Http1ResponseCodeDetails::get().ContentLengthNotAllowed);
+      throw CodecProtocolException(
+          "http/1.1 protocol error: both 'Content-Length' and 'Transfer-Encoding' are set.");
```
cc @asraa 

Some changes required here after merge with #11778 which removes exception use from the non-legacy http1 codec_impl.cc

antoniovicente commented on test/common/http/http1/codec_impl_test.cc:
```diff
@@ -2826,5 +2886,29 @@ TEST_P(Http1ClientConnectionImplTest, ManyResponseHeadersAccepted) {
+TEST_P(Http1ServerConnectionImplTest, TestSmugglingAllowChunkedContentLength100) {
+  // content-length greater than POST body size
+  testAllowChunkedContentLength(100, true);
+}
```
I'm not seeing similar tests for the response side which operate on Http1ClientConnectionImpl.

veshij replied:
Found an issue with http parser in response test case, will updated PR when it's fixed.
/wait

veshij replied:
added tests for response case

veshij commented on source/common/http/http1/codec_impl.h:
```diff
@@ -241,6 +241,7 @@ class ConnectionImpl : public virtual Connection, protected Logger::Loggable<Log
 
   Network::Connection& connection_;
   CodecStats& stats_;
+  const Http1Settings codec_settings_;
```
I'm not quite sure about `settings` lifetime, could it be reference?

mattklein123 replied:
I mean pass as a reference parameter, not store as a reference, otherwise you may copy twice. I think it may be safe to actually store as a reference but that would need to be analyzed and it is probably not worth it if the object is small.

mattklein123 commented:
Note will also need a main merge.

veshij commented:
Found an issue with http-parser for response parsing case, will update PR when issue is fixed.

alyssawilk commented on source/common/http/http1/codec_impl_legacy.cc:
```diff
@@ -679,6 +679,29 @@ int ConnectionImpl::onHeadersCompleteBase() {
+      request_or_response_headers.removeContentLength();
+    } else {
+      error_code_ = Http::Code::BadRequest;
+      sendProtocolError(Http1ResponseCodeDetails::get().ContentLengthNotAllowed);
```
preference for a new detail field - generally we want the rc detail field to indicate the block of code which caused Envoy to reject a request - if we're using one twice it then becomes ambiguous.  C

alyssawilk commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -730,6 +730,29 @@ Envoy::StatusOr<int> ConnectionImpl::onHeadersCompleteBase() {
+    if ((parser_.flags & F_CHUNKED) && codec_settings_.allow_chunked_length_) {
+      request_or_response_headers.removeContentLength();
+    } else {
+      error_code_ = Http::Code::BadRequest;
```
Do we have an integration test that covers this?    If not I'd like to make sure we add one.

veshij replied:
Added server/client integration tests for enabled allow_chunked_length_.
Existing smuggling test covers disabled allow_chunked_length_ logic.

antoniovicente commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -1086,7 +1086,8 @@ bool ClientConnectionImpl::cannotHaveBody() {
   } else if (parser_.status_code == 204 || parser_.status_code == 304 ||
-             (parser_.status_code >= 200 && parser_.content_length == 0)) {
+             (parser_.status_code >= 200 && parser_.content_length == 0 &&
+              !(parser_.flags & F_CHUNKED))) {
```
Out of curiosity: what coverage do we have for 1xx responses that include content-length or transfer-encoding: chunked headers?

veshij commented:
/wait

ceastman-ibm commented:
When would this PR get into envoyproxy/envoy-wasm so that istio can pick it up? We are actually hitting this edge case. @alyssawilk

alyssawilk commented:
Up to folks on that side.
Did @jplevyak 's last merge sort this out for you?  Looks like it landed the same day you pinged here.

ceastman-ibm commented:
@alyssawilk we won't be able to test this until this version of envoyproxy gets into istio and then into our IBM managed istio offering.
